activerecord/lib/active_record/connection_adapters/abstract/query_log_tags.rb

```diff
@@ -0,0 +1,205 @@
+        thread_mattr_accessor :cached_comment, instance_accessor: false
+
+        class << self
+          # Updates the context used to construct the query log tags.
+          # Resets the cached comment if <tt>cache_query_log_tags</tt> is +true+.
+          def update(**options)
+            context.merge!(**options.symbolize_keys)
+            self.cached_comment = nil
+          end
+
+          def tags_available?
+            components.present? || inline_annotations.present?
+          end
+
+          # Returns an SQL comment +String+ containing the query log tags.
+          # Sets and returns a cached comment if <tt>cache_query_log_tags</tt> is +true+.
+          def comment
+            if cache_query_log_tags
+              self.cached_comment ||= uncached_comment
+            else
+              uncached_comment
+            end
+          end
+
+          def uncached_comment
+            content = tag_content
+            if content.present?
+              "/*#{escape_sql_comment(content)}*/"
+            else
+              ""
+            end
+          end
+
+          # Returns a +String+ containing any inline comments from +with_annotation+.
+          def inline_comment
+            return nil unless inline_annotations.present?
+            "/*#{escape_sql_comment(inline_tag_content)}*/"
+          end
+
+          # Manually clear the comment cache.
+          def clear_comment_cache!
+            self.cached_comment = nil
+          end
+
+          # Annotate any query within `&block`. Can be nested.
+          def with_annotation(comment, &block)
+            self.inline_annotations.push(comment)
+            block.call if block.present?
+          ensure
+            self.inline_annotations.pop
+          end
+
+          # Return the set of active inline annotations from +with_annotation+.
+          def inline_annotations
+            context[:inline_annotations] ||= []
+          end
+
+          # QueryLogTags +component+ methods
+
+          # Set during Rails boot in lib/active_record/railtie.rb
+          def application # :nodoc:
+            context[:application_name]
+          end
+
+          def pid # :nodoc:
+            Process.pid
+          end
+
+          def connection_config # :nodoc:
+            ActiveRecord::Base.connection_db_config
+          end
+
+          def socket # :nodoc:
+            connection_config.socket
+          end
+
+          def db_host # :nodoc:
+            connection_config.host
+          end
+
+          def database # :nodoc:
+            connection_config.database
+          end
+
+          def line # :nodoc:
+            backtrace_cleaner.add_silencer { |line| line.match?(/lib\/active_(record|support)/) }
+            backtrace_cleaner.clean(caller.lazy).first
+          end
+
+          private
+            def context
+              Thread.current[:active_record_query_log_tags_context] ||= {}
```

Comment:
I think we need to clear the context at the end of every request/job, otherwise it's too easy for it to leak:

```ruby
before_action, if -> { current_user.admin? } do
  ActiveRecord::ConnectionAdapters::AbstractAdapter::QueryLogTagsContext.update(admin: true)
end
```

This should probably be implemented as an [executor hook](https://github.com/rails/rails/blob/7612243b889169eb3fde0733e0b84a9f66113122/activesupport/lib/active_support/execution_wrapper.rb#L50).

Reply:
I'll look into executor hooks, thanks for the tip - may be a much cleaner approach to wrapping everything in action/perform callback filters (taken from Marginalia approach), which currently reset the context and wipe the cache around each request - confirmed in tests I've just added.

Reply:
At the moment we're clearing the controller/job-specific entries in the context much like Marginalia does, but I can change this to clear the whole context.

When I originally replied I wasn't really sure what executor hooks were :) I'm not attuned to the concurrency implications of this behaviour just yet, I'll read up some more - any pointers appreciated though!

Reply:
Resolving as this behaviour has been changed to ensure the context provided to a block update is reset.